Userland/Libraries/LibProtocol/WebSocket.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021, The SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Badge.h>
+#include <AK/ByteBuffer.h>
+#include <AK/Function.h>
+#include <AK/RefCounted.h>
+#include <AK/String.h>
+#include <AK/WeakPtr.h>
+#include <LibCore/Notifier.h>
+#include <LibIPC/Forward.h>
+
+namespace Protocol {
+
+class WebSocketClient;
+
+class WebSocket : public RefCounted<WebSocket> {
+public:
+    struct CertificateAndKey {
+        String certificate;
+        String key;
+    };
+
+    struct Message {
+        ByteBuffer data;
+        bool is_text;
+    };
+
+    enum class Error {
+        CouldNotEstablishConnection,
+        ConnectionUpgradeFailed,
+        ServerClosedSocket,
+    };
+
+    enum class ReadyState {
+        Connecting = 0,
+        Open = 1,
+        Closing = 2,
+        Closed = 3,
+    };
+
+    static NonnullRefPtr<WebSocket> create_from_id(Badge<WebSocketClient>, WebSocketClient& client, i32 connection_id)
+    {
+        return adopt_ref(*new WebSocket(client, connection_id));
+    }
+
+    int id() const { return m_connection_id; }
+
+    ReadyState ready_state();
+
+    void send(ByteBuffer, bool is_text);
+    void send(StringView);
```

Comment:
Would `ReadonlyBytes` make more sense (semantically) than `StringView` here?
It's arbitrary data right?

Reply:
The idea is to have two options (text or binary data). The WebSocket protocol has a flag in the frame that indicates whether the data is UTF-8 or not, which is what the first method allows to set explicitely and the second method always set implicitely to "true".

Reply:
Gotcha :)

Reply:
I added parameter names to make that clearer:
```C++
void send(ByteBuffer binary_or_text_message, bool is_text);
void send(StringView text_message);
```